### PR TITLE
Add OneAgent metadata enrichment and improve example

### PR DESCRIPTION
**Pull request by pirgeo**

### CODE REVIEW

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        for key, value in self._parse_oneagent_metadata(
+                self.__get_metadata_file_content()).items():
+            tags[key] = value
+
+    def __get_metadata_file_content(self) -> List[str]:
+        try:
+            metadata_file_name = None
+            with open(
+                    "dt_metadata_e617c525669e072eebe3d0f08212e8f2.properties",
+                    "r") as magic_file:
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:35]:
```suggestion
                    "r") as metadata_indirection_file:
```
maybe?

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                metadata_file_name = magic_file.read()
+
+            if not metadata_file_name:
+                self.__logger.debug("metadata file name not specified by "
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:39]: logger can be None

Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:39]: This should be at least a warning since this would likely be a bug in the agent, right?

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                                    "OneAgent.")
+                return []
+
+            with open(metadata_file_name, "r") as attributes_file:
+                return attributes_file.readlines()
+
+        except OSError:
+            logging.warning(
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:47]: Are you using the root logger on purpose?
I would also make this an info since this might be expected.

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                "could not read OneAgent metadata file. This is normal if "
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:48]:
```suggestion
                "Could not read OneAgent metadata file. This is normal if "
```

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                "OneAgent is not installed.")
+
+    def _parse_oneagent_metadata(self, lines) -> Mapping[str, str]:
+        key_value_pairs = {}
+        for line in lines:
+            self.__logger.debug("parsing line {}".format(line))
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:54]: Maybe prefix all log lines with `[OneAgentMetadataEnricher]` as you're passing the same logger used elsewhere in the exporter.

File: test/test_oneagentmetadataenricher.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+class TestOneAgentMetadataEnricher(unittest.TestCase):
+    __logger = logging.Logger(__name__)
+
+    def test_parse_oneagent_metadata(self):
+        enricher = OneAgentMetadataEnricher(self.__logger)
+        parsed = enricher._parse_oneagent_metadata(["key1=value1",
+                                                    "key2=value2"])
+
+        self.assertEqual("value1", parsed["key1"])
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:29]: Add an assertion that there are only two entries.

File: test/test_oneagentmetadataenricher.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    def test_parse_invalid_metadata(self):
+        enricher = OneAgentMetadataEnricher(self.__logger)
+
+        self.assertFalse(enricher._parse_oneagent_metadata(
+            ["=0x5c14d9a68d569861"]))
+        self.assertFalse(enricher._parse_oneagent_metadata(["otherKey="]))
+        self.assertFalse(enricher._parse_oneagent_metadata([""]))
+        self.assertFalse(enricher._parse_oneagent_metadata(["="]))
+        self.assertFalse(enricher._parse_oneagent_metadata(["==="]))
+        self.assertFalse(enricher._parse_oneagent_metadata([]))
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:41]: Add a test that valid + invalid lines combined still return the valid ones.

File: test/test_oneagentmetadataenricher.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import unittest
+from unittest.mock import patch
+
+from dynatrace.opentelemetry.metrics.export import OneAgentMetadataEnricher
+
+
+class TestOneAgentMetadataEnricher(unittest.TestCase):
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:21]: Please add tests for the file handling (redirection file not being present or empty, containing nothing or an invalid file name, actual file not being present or empty, containing invalid or empty data).

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py (hunk 3 above)
Comment (pirgeo) [reply src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:47]: changed it to an enricher-local logger

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py (hunk 2 above)
Comment (pirgeo) [reply src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:39]: now automatically set up in the constructor, should no longer be able to be None

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py (hunk 5 above)
Comment (pirgeo) [reply src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:54]: Now using a OneAgentMetadataEnricher-local logger

File: test/test_oneagentmetadataenricher.py (hunk 8 above)
Comment (pirgeo) [reply test/test_oneagentmetadataenricher.py:22]: added a bunch of tests

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py [hunk 9]
```diff
@@ -19,51 +19,65 @@
...
 
             with open(metadata_file_name, "r") as attributes_file:
                 return attributes_file.readlines()
-
         except OSError:
-            logging.warning(
-                "could not read OneAgent metadata file. This is normal if "
-                "OneAgent is not installed.")
+            self.__logger.info(
+                "Could not read OneAgent metadata file.")
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:61]: You could print the file path here so users could check the permissions on the file.

File: test/test_oneagentmetadataenricher.py [hunk 10]
```diff
@@ -58,10 +75,122 @@ def test_mock_get_metadata_file(self, mock_func):
...
         self.assertEqual(tags, {"tag1": "newValue"})
+
+    @patch('dynatrace.opentelemetry.metrics.export.oneagentmetadataenricher'
+           '.OneAgentMetadataEnricher._get_metadata_file_content')
+    def test_parse_valid_data(self, mock_func):
+        mock_func.return_value = ["k1=v1\n", "k2=v2"]
+
+        enricher = OneAgentMetadataEnricher()
+        res = enricher._parse_oneagent_metadata(
+            enricher._get_metadata_file_content())
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:93]: Since you pass the values anyway you could omit mocking the method and just pass it right away.

File: test/test_oneagentmetadataenricher.py [hunk 11]
```diff
@@ -58,10 +75,122 @@ def test_mock_get_metadata_file(self, mock_func):
...
+        self.assertEqual("v2", res["k2"])
+
+    @patch('dynatrace.opentelemetry.metrics.export.oneagentmetadataenricher'
+           '.OneAgentMetadataEnricher._get_metadata_file_content')
+    def test_parse_empty_list(self, mock_func):
+        mock_func.return_value = []
+
+        enricher = OneAgentMetadataEnricher()
+        res = enricher._parse_oneagent_metadata(
+            enricher._get_metadata_file_content())
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:106]: Same as above.

File: test/test_oneagentmetadataenricher.py [hunk 12]
```diff
@@ -58,10 +75,122 @@ def test_mock_get_metadata_file(self, mock_func):
...
+
+        with patch("builtins.open", mock):
+            res = enricher._get_metadata_file_content()
+            self.assertEqual(0, len(res))
+
+    def test_get_file_contents_success(self):
+        mock = mock_open(read_data="key1=value1\nkey2=value2")
+        enricher = OneAgentMetadataEnricher()
+        with patch("builtins.open", mock):
+            res = enricher._get_metadata_file_content()
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:149]: This test is not quite valid since it will return `"key1=value1\nkey2=value2"` as file name on the first open, try to open this file name and it will succeed because the second open call is also mocked.
I think `test_get_file_contents_from_tmpfile` below, which tests it properly, is sufficient, so you can remove this misleading test case here or add a comment that this only covers `_get_metadata_file_content` partially.

File: test/test_oneagentmetadataenricher.py [hunk 13]
```diff
@@ -58,10 +75,122 @@ def test_mock_get_metadata_file(self, mock_func):
...
+        mock_func.return_value = None
+
+        enricher = OneAgentMetadataEnricher()
+        res = enricher._get_metadata_file_content()
+        self.assertEqual(0, len(res))
+
+    @patch('dynatrace.opentelemetry.metrics.export.oneagentmetadataenricher'
+           '.OneAgentMetadataEnricher._get_metadata_file_name')
+    def test_get_file_contents_from_tmpfile(self, mock_func):
+        with tempfile.TemporaryDirectory() as tmp_dir:
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:164]: Does this clean up the directory afterwards or will you have to delete the file manually?

File: test/test_oneagentmetadataenricher.py [hunk 14]
```diff
@@ -58,10 +75,122 @@ def test_mock_get_metadata_file(self, mock_func):
...
+            tmpfile_name = tmp_dir + "/tmp_file"
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:165]: Does appending like this work cross-platform (`\` on Windows)?

File: example/basic_sample.py [hunk 15]
```diff
@@ -17,49 +17,93 @@
...
 
-endpoint_url = None
-api_token = None
-exporter = DynatraceMetricsExporter(endpoint_url, api_token)
+def parse_arguments():
+    parser = argparse.ArgumentParser(
+        description="Example exporting metrics using the Dynatrace metrics "
+                    "exporter.",
+        epilog="The script can be run without any arguments. In that case, "
+               "the local OneAgent is used as an endpoint.")
```
Comment (arminru) [inline example/basic_sample.py:28]:
```suggestion
               "the local OneAgent is used as an endpoint, if installed.")
```

File: example/basic_sample.py [hunk 16]
```diff
@@ -17,49 +17,93 @@
...
+    parser.add_argument("-e", "--endpoint", type=str, dest="endpoint",
+                        default="http://127.0.0.1:14499/metrics/ingest",
+                        help="The endpoint url used to export metrics to. "
+                             "This can be either a dynatrace metrics "
```
Comment (arminru) [inline example/basic_sample.py:32]:
```suggestion
                             "This can be either a Dynatrace metrics "
```

File: example/basic_sample.py [hunk 17]
```diff
@@ -17,49 +17,93 @@
...
+                             "ingestion endpoint, or a local OneAgent "
+                             "endpoint. If no value is set, use the default "
+                             "local OneAgent endpoint")
```
Comment (arminru) [inline example/basic_sample.py:35]:
```suggestion
                             "endpoint. If no value is set, the default "
                             "local OneAgent endpoint is used.")
```

File: example/basic_sample.py [hunk 18]
```diff
@@ -17,49 +17,93 @@
...
-    description="number of requests",
-    unit="1",
-    value_type=int
-)
+    parser.add_argument("-nm", "--no-metadata", dest="metadata_enrichment",
+                        action="store_false",
+                        help="Turn off OneAgent Metadata enrichment. If no "
+                             "OneAgent is running on the machine, this is "
+                             "ignored. Otherwise, OneAgent metadata will be "
+                             "added to each of the exported metric lines")
```
Comment (arminru) [inline example/basic_sample.py:48]:
```suggestion
                             "added to each of the exported metric lines.")
```

File: example/basic_sample.py [hunk 19]
```diff
@@ -17,49 +17,93 @@
...
+if __name__ == '__main__':
+    args = parse_arguments()
 
-# Update the metric instruments using the direct calling convention
-requests_counter.add(25, staging_labels)
-requests_size.record(100, staging_labels)
-time.sleep(10)
+    # set up opentelemetry for export:
+    metrics.set_meter_provider(MeterProvider())
+    meter = metrics.get_meter(__file__)
```
Comment (arminru) [inline example/basic_sample.py:59]: Shouldn't this be `__name__`?

File: example/basic_sample.py [hunk 20]
```diff
@@ -17,49 +17,93 @@
...
 
-requests_counter.add(50, staging_labels)
-requests_size.record(5000, staging_labels)
-time.sleep(5)
+    exporter = DynatraceMetricsExporter(args.endpoint, args.token,
+                                        export_oneagent_metadata=
+                                        args.metadata_enrichment)
```
Comment (arminru) [inline example/basic_sample.py:63]: odd line break :)

File: example/basic_sample.py [hunk 21]
```diff
@@ -17,49 +17,93 @@
...
 
-requests_counter.add(35, testing_labels)
-requests_size.record(2, testing_labels)
+    # this line registers the meter and exporter with the MeterProvider set
```
Comment (arminru) [inline example/basic_sample.py:65]:
```suggestion
    # This call registers the meter and exporter with the global MeterProvider set
```

File: test/test_oneagentmetadataenricher.py (hunk 13 above)
Comment (pirgeo) [reply test/test_oneagentmetadataenricher.py:164]: "On completion of the context or destruction of the temporary directory object the newly created temporary directory and all its contents are removed from the filesystem.", so it should be gone.

File: example/basic_sample.py [hunk 22]
```diff
@@ -17,49 +17,93 @@
...
+    # above. All instruments created by the meter that is registered here will
+    # export to the Dynatrace metrics exporter. It is a good idea to keep a 
+    # reference to the meter (e. g. in a global variable) in order to create
+    # instruments anywhere in the code that all export to the same Dynatrace
+    # metrics exporter.
+    metrics.get_meter_provider().start_pipeline(meter, exporter, 5.)
```
Comment (arminru) [inline example/basic_sample.py:71]: You could also make the export interval configurable. I'd suggest 10-60 secs in the arg parser and would make 10 secs the default.

File: test/test_oneagentmetadataenricher.py (hunk 14 above)
Comment (pirgeo) [reply test/test_oneagentmetadataenricher.py:165]: probably not, switched to `os.path.join`

File: example/basic_sample.py [hunk 23]
```diff
@@ -17,49 +17,93 @@
...
-requests_size.record(100, staging_labels)
-time.sleep(10)
+    # set up opentelemetry for export:
+    metrics.set_meter_provider(MeterProvider())
+    meter = metrics.get_meter(__file__)
 
-requests_counter.add(50, staging_labels)
-requests_size.record(5000, staging_labels)
-time.sleep(5)
+    exporter = DynatraceMetricsExporter(args.endpoint, args.token,
```
Comment (arminru) [inline example/basic_sample.py:61]: Please add `otel.python` as a prefix to make it easier to find it in the UI. The current metric names "requests" and "request_size" are quite generic.

File: example/basic_sample.py [hunk 24]
```diff
@@ -17,49 +17,93 @@
...
+        name="requests",
+        description="number of requests",
+        unit="1",
+        value_type=int
+    )
+
+    requests_size = meter.create_valuerecorder(
+        name="request_size",
+        description="size of requests",
+        unit="1",
```
Comment (arminru) [inline example/basic_sample.py:83]:
```suggestion
        name="request_size_bytes",
        description="size of requests",
        unit="byte",
```

File: test/test_oneagentmetadataenricher.py (hunk 12 above)
Comment (pirgeo) [reply test/test_oneagentmetadataenricher.py:149]: oh, that's true! updated it so it will return `indirection_filename` on the first call. also added the info that it only partially tests the function

File: test/test_oneagentmetadataenricher.py (hunk 10 above)
Comment (pirgeo) [reply test/test_oneagentmetadataenricher.py:93]: :sweat_smile:

File: example/basic_sample.py (hunk 19 above)
Comment (pirgeo) [reply example/basic_sample.py:59]: `__name__` will return `__main__` in this case which is probably not helpful. i just checked again, `__file__` will add the whole path + extension, so i changed it so it only uses the file name (no path, no ext)

File: example/basic_sample.py [hunk 25]
```diff
@@ -25,14 +26,14 @@ def parse_arguments():
...
         epilog="The script can be run without any arguments. In that case, "
-               "the local OneAgent is used as an endpoint.")
+               "the local OneAgent is used as an endpoint, if it is installed.")
     parser.add_argument("-e", "--endpoint", type=str, dest="endpoint",
                         default="http://127.0.0.1:14499/metrics/ingest",
                         help="The endpoint url used to export metrics to. "
-                             "This can be either a dynatrace metrics "
+                             "This can be either a Dynatrace metrics "
                              "ingestion endpoint, or a local OneAgent "
                              "endpoint. If no value is set, use the default "
```
Comment (arminru) [inline example/basic_sample.py:35]:
```suggestion
                             "endpoint. If no value is set, the default "
```

File: example/basic_sample.py [hunk 26]
```diff
@@ -45,7 +46,15 @@ def parse_arguments():
                         help="Turn off OneAgent Metadata enrichment. If no "
                              "OneAgent is running on the machine, this is "
                              "ignored. Otherwise, OneAgent metadata will be "
-                             "added to each of the exported metric lines")
+                             "added to each of the exported metric lines.")
+
+    parser.add_argument("-i", "-interval", default=10., type=float,
```
Comment (arminru) [inline example/basic_sample.py:51]:
```suggestion
    parser.add_argument("-i", "--interval", default=10., type=float,
```

File: example/basic_sample.py [hunk 27]
```diff
@@ -45,7 +46,15 @@ def parse_arguments():
...
+                        dest="interval",
+                        help="Set the export interval for the Dynatrace metrics"
+                             " exporter. This specifies how often data is "
```
Comment (arminru) [inline example/basic_sample.py:54]:
```suggestion
                             " exporter in seconds. This specifies how often data is "
```

File: src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py [hunk 28]
```diff
@@ -58,7 +58,8 @@ def _get_metadata_file_content(self) -> List[str]:
                 return attributes_file.readlines()
         except OSError:
             self.__logger.info(
-                "Could not read OneAgent metadata file.")
+                "Could not read OneAgent metadata file. ({})".format(
```
Comment (arminru) [inline src/dynatrace/opentelemetry/metrics/export/oneagentmetadataenricher.py:61]:
```suggestion
                "Could not read OneAgent metadata file ({}).".format(
```

File: test/test_oneagentmetadataenricher.py [hunk 29]
```diff
@@ -143,7 +136,16 @@ def test_indirection_file_empty(self):
...
 
     def test_get_file_contents_success(self):
-        mock = mock_open(read_data="key1=value1\nkey2=value2")
+        # this only tests _get_metadata_file_content() paritally since both
+        # open calls are mocked, so even if the file does not exist this test
+        # will pass since the second open call is also mocked.
+        mock = mock_open()
+        # return "indirection_filename" on the first open call and the
+        # key-vlaue pairs on the second.
+        mock.side_effect = [mock_open(read_data=x).return_value for x in
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:145]: Is the side effect evaluated before or after it returns `read_data`?

File: test/test_oneagentmetadataenricher.py [hunk 30]
```diff
@@ -143,7 +136,16 @@ def test_indirection_file_empty(self):
...
+                            ["indirection_filename",
+                             "key1=value1\nkey2=value2"]]
```
Comment (arminru) [inline test/test_oneagentmetadataenricher.py:147]: Please also verify the content here and in `test_get_file_contents_from_tmpfile` in addition to the length.

File: test/test_oneagentmetadataenricher.py (hunk 29 above)
Comment (pirgeo) [reply test/test_oneagentmetadataenricher.py:145]: the object named `mock` does not have a `read_data` property in this case. therefore, on the first `open` call, `mock_open(read_data="indirection_filename").return_value` is returned and on the second `open` call `mock_open(read_data="key1=value1\nkey2=value2").return_value` is returned, so the object called `mock` is like a wrapper for the other two.

**Review by arminru — APPROVED** [review]
LGTM - Thanks!
